### ISSUE TITLE
possible race condition in get_devname()

### ISSUE THREAD

**Issue by jtluka** [jpirko/lnst#181]
I hit following traceback while running one of my tests.

```
    Traceback (most recent call last):
      File "/usr/bin/lnst-ctl", line 182, in get_recipe_result
        res = exec_action(action, nettestctl)
      File "/usr/bin/lnst-ctl", line 137, in exec_action
        return nettestctl.run_recipe()
      File "/usr/lib/python2.7/site-packages/lnst/Controller/NetTestController.py", line 667, in run_recipe
        res = self._run_recipe()
      File "/usr/lib/python2.7/site-packages/lnst/Controller/NetTestController.py", line 685, in _run_recipe
        res = self._run_task(task)
      File "/usr/lib/python2.7/site-packages/lnst/Controller/NetTestController.py", line 733, in _run_task
        return self._run_python_task(task)
      File "/usr/lib/python2.7/site-packages/lnst/Controller/NetTestController.py", line 722, in _run_python_task
        module = imp.load_source(name, task_path.resolve())
      File "/mnt/testarea/tunnel-lnst-tests/gre/gre.py", line 96, in <module>
        m1.run("ethtool -K %s %s " % (dev, dev_features))
      File "/usr/lib/python2.7/site-packages/lnst/Controller/Task.py", line 850, in __str__
        return str(self.get_val())
      File "/usr/lib/python2.7/site-packages/lnst/Controller/Task.py", line 847, in get_val
        return self._func(*self._args, **self._kwargs)
      File "/usr/lib/python2.7/site-packages/lnst/Controller/Machine.py", line 662, in get_devname
        raise MachineError(msg)
    MachineError: Device name is not available for interface 'gen_eth_0'
```

The device name returned by get_devname() is a generic name generated by LNST Controller before it gets update from LNST Slave with real name. I believe this is a race condition.

The recipe and python task:

```xml
<lnstrecipe>
    <network>
        <host id="testmachine1">
            <params/>
            <interfaces>
                <eth id="nic1" label="testnet">
                    <addresses>
                        <address>192.168.100.2/24</address>
                        <address>fe01::1/64</address>
                    </addresses>
                </eth>
            </interfaces>
        </host>
        <host id="testmachine2">
            <params/>
            <interfaces>
                <eth id="nic1" label="testnet">
                    <addresses>
                        <address>192.168.100.3/24</address>
                        <address>fe01::2/64</address>
                    </addresses>
                </eth>
            </interfaces>
        </host>
    </network>

    <!--
    Possible tests are:
        icmp
        icmp6
        igmp
        perf
        perf6
    -->
    <define>
        <alias name="tests" value=""/>
    </define>

    <task python="gre.py"/>
</lnstrecipe>
```

```python
from lnst.Controller.Task import ctl


test_list = ctl.get_alias('tests')
test_set = []
if test_list != None:
    for t in test_list.split(','):
        if len(t) > 0:
            test_set.append(t)

if len(test_set) == 0:
    test_set=['icmp', 'icmp6', 'igmp', 'perf', 'perf6']

gre_devname = "gre1"

m1 = ctl.get_host("testmachine1")
m2 = ctl.get_host("testmachine2")

m1_tif = m1.get_interface("nic1")
m2_tif = m2.get_interface("nic1")

m1_nics = []
m2_nics = []

for k, iface in m1.get_interfaces().iteritems():
    if iface.get_type() == 'eth':
        m1_nics.append(iface.get_devname())
for k, iface in m2.get_interfaces().iteritems():
    if iface.get_type() == 'eth':
        m2_nics.append(iface.get_devname())

# I removed the rest of the python task since the error happens here
```

Originally this happened on baremetal machines. I was not able to reproduce this in virtual environment.

**Comment by jtluka** [comment]
The python task should be:
```python
from lnst.Controller.Task import ctl


test_list = ctl.get_alias('tests')
test_set = []
if test_list != None:
    for t in test_list.split(','):
        if len(t) > 0:
            test_set.append(t)

if len(test_set) == 0:
    test_set=['icmp', 'icmp6', 'igmp', 'perf', 'perf6']

gre_devname = "gre1"

m1 = ctl.get_host("testmachine1")
m2 = ctl.get_host("testmachine2")

m1_tif = m1.get_interface("nic1")
m2_tif = m2.get_interface("nic1")

m1_nics = []
m2_nics = []

for k, iface in m1.get_interfaces().iteritems():
    if iface.get_type() == 'eth':
        m1_nics.append(iface.get_devname())
for k, iface in m2.get_interfaces().iteritems():
    if iface.get_type() == 'eth':
        m2_nics.append(iface.get_devname())

m1.sync_resources(modules=["PacketAssert",
                           "IcmpPing",
                           "Icmp6Ping",
                           "Multicast",
                           "Iperf"],
                  tools=["multicast"])

m2.sync_resources(modules=["Multicast", "Iperf"], tools=["multicast"])

a_end = m1.get_ip("nic1",0)
b_end = m2.get_ip("nic1",0)

a_ip4 = "192.168.6.2/24"
a_net4 = "192.168.6.0/24"
b_ip4 = "192.168.7.2/24"
b_net4 = "192.168.7.0/24"

a_ip6 = "6001:db8:ac10:fe01::2/64"
a_net6 = "6001:db8:ac10:fe01::0/64"
b_ip6 = "7001:db8:ac10:fe01::2/64"
b_net6 = "7001:db8:ac10:fe01::0/64"

nic1_name = m1.get_devname("nic1")

m1.run("ip tunnel add %s mode gre remote %s local %s ttl 64" % (gre_devname,
    b_end, a_end))

m1.run("ip link set dev %s up" % gre_devname)
m1.run("ip -4 a add %s dev %s" % (a_ip4, gre_devname))
m1.run("ip -6 a add %s dev %s" % (a_ip6, gre_devname))
m1.run("ip -4 route add %s dev %s" % (b_net4, gre_devname))
m1.run("ip -6 route add %s dev %s" % (b_net6, gre_devname))

m1.run("ip route")

m2.run("ip tunnel add %s mode gre remote %s local %s ttl 64" % (gre_devname,
    a_end, b_end))

m2.run("ip link set dev %s up" % gre_devname)
m2.run("ip -4 a add %s dev %s" % (b_ip4, gre_devname))
m2.run("ip -6 a add %s dev %s" % (b_ip6, gre_devname))
m2.run("ip -4 route add %s dev %s" % (a_net4, gre_devname))
m2.run("ip -6 route add %s dev %s" % (a_net6, gre_devname))

m2.run("ip route")

ctl.wait(seconds=30)

offload_settings = [
                    [("gro", "on"), ("gso", "on"), ("tso", "on")],
                    [("gro", "off"), ("gso", "on"), ("tso", "on")],
                    [("gro", "on"), ("gso", "off"), ("tso", "on")],
                    [("gro", "on"), ("gso", "on"), ("tso", "off")],
                   ]

offloads=['gro','gso','tso']

for settings in offload_settings:
    dev_features = ""
    for offload in settings:
        dev_features += " %s %s" % (offload[0], offload[1])
    for dev in m1_nics:
        m1.run("ethtool -K %s %s " % (dev, dev_features))
    for dev in m2_nics:
        m2.run("ethtool -K %s %s " % (dev, dev_features))
```

**Comment by olichtne** [comment]
[Patch](https://lists.fedorahosted.org/archives/list/lnst-developers@lists.fedorahosted.org/message/SLNVPQJSPO66F2D4MVIVPMTYZIFFJIAS/) posted to the mailing list, please try it out.

**Comment by olichtne** [comment]
[v2](https://lists.fedorahosted.org/archives/list/lnst-developers@lists.fedorahosted.org/message/2WV6FUXJ3NVNDDUX2L5Q4UWWK2QSUGC5/) sent, based on your comment.

**Comment by jtluka** [comment]
The patch fixed the issue. Thanks!

Fixed in e79432a43669aee0af126ef6d2757a6204659747
